Feature/custom buttons remote

nikos1988 opened the pull request:
PR 3/3

nikos1988 committed:
custom buttons implementation

nikos1988 committed:
do not close button container

nikos1988 committed:
added MQTT integration schema

martonborzak commented:
@nikos1988 could you please write a brief description of what the PR does? Thanks!

martonborzak reviewed (changes requested):
An icon font is used to store icons. It would be better to stick to that and expand it, rather than bringing in SVGs. I can help with this @nikos1988

martonborzak commented on components/remote/ui/ir/Card.qml:
```diff
@@ -68,52 +68,164 @@ Rectangle {
     }
 
+    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+    // BUTTON REPEAT TIMER
```
Instead of all these timers, I think a component could be use that could handle repeated button presses. Similar solution to the volume control: https://github.com/YIO-Remote/remote-software/blob/master/basic_ui/Volume.qml

nikos1988 replied:
Yes, that can be done. I will look into it

nikos1988 replied:
The PR makes some changes to the remote entity and adds the possibility for integrations to add custom features (buttons) to entities. The idea was to be able to replicate all functions of a remote, e.g. AVR or TV where the supported features are just not enough.

The custom buttons are shown in an extra page on the swipeview of the remote entity. I also added a page for the colored function buttons. Furthermore i added icons for the transport feature buttons and did some minor cosmetic changes.

nikos1988 commented:
> 
> 
> An icon font is used to store icons. It would be better to stick to that and expand it, rather than bringing in SVGs. I can help with this @nikos1988

Yes that would be nice

nikos1988 commented:
> 
> 
> @nikos1988 could you please write a brief description of what the PR does? Thanks!



The PR makes some changes to the remote entity and adds the possibility for integrations to add custom features (buttons) to entities. The idea was to be able to replicate all functions of a remote, e.g. AVR or TV where the supported features are just not enough.

The custom buttons are shown in an extra page on the swipeview of the remote entity. I also added a page for the colored function buttons. Furthermore i added icons for the transport feature buttons and did some minor cosmetic changes.

nikos1988 committed:
removed sendCustomCommand, use sendCommand with param "custom_command" instead

nikos1988 committed:
updated mqtt ip title and example

zehnm commented on config-schema.json:
```diff
@@ -545,6 +560,92 @@
+                    "default": null,
+                    "examples": [
+                      "MQTT"
+                    ],
+                    "pattern": ""
+                  },
+                  "id": {
+                    "$id": "#/properties/integrations/properties/mqtt/properties/data/items/properties/id",
+                    "type": "string",
+                    "title": "The Id Schema",
+                    "default": null,
+                    "examples": [
+                      "mqtt"
+                    ],
+                    "pattern": ""
+                  }
+                }
+              }
+            },
+            "log": {
+              "$id": "#/properties/integrations/properties/mqtt/properties/log",
+              "type": "string",
+              "title": "The Log Schema",
+              "default": null,
+              "examples": [
+                "debug"
+              ],
+              "pattern": ""
+            },
+            "mdns": {
+              "$id": "#/properties/integrations/properties/mqtt/properties/mdns",
+              "type": "string",
+              "title": "The Mdns Schema",
+              "examples": [
+                "_mqtt._tcp"
+              ],
+              "pattern": "^(.*)$"
+            }
+          }
+        },
```
MQTT is neither part of remote-software nor this pull request.
Please remove.

nikos1988 replied:
Ok, this has been removed

zehnm commented on sources/entities/entities.cpp:
```diff
@@ -77,6 +77,9 @@ void Entities::load() {
                 QVariantMap           map = type[k].toMap();
                 QObject *             obj = Integrations::getInstance()->get(map.value("integration").toString());
                 IntegrationInterface *integration = qobject_cast<IntegrationInterface *>(obj);
+                if (!integration) {
+                    qCWarning(CLASS_LC()) << "integration pointer is null" << map.value("integration").toString();
+                }
```
If integration is null we should probably not add it to m_supportedEntities which will most likely produce a runtime error later on.
@martonborzak or is there anything I'm missing? I.e. can an entity have no integration?

martonborzak replied:
Entities usually have integrations. Null pointer will crash the app.

zehnm commented on sources/yioapi.cpp:
```diff
@@ -166,6 +166,10 @@ bool YioAPI::addEntity(QVariantMap entity) {
         if (obj) {
             IntegrationInterface *integration = qobject_cast<IntegrationInterface *>(obj);
 
+            if (!integration) {
+                qCWarning(CLASS_LC) << "integration pointer is null" << entity.value(Config::KEY_ENTITY_ID);
+            }
```
See above. If an entity must have an integration, then it shouldn't be added to m_entities.

nikos1988 committed:
removed mqtt integration schema

nikos1988 committed:
removed unnecessary logging

nikos1988 commented:
I removed the unnecessary logging

nikos1988 committed:
do not add empty views (no buttons) to remote entity ui swipeview

nikos1988 committed:
use function to determine pages to load

nikos1988 committed:
use Components with ids for swipeview pages

nikos1988 committed:
removed some debug logging

nikos1988 committed:
set view visible after loading

nikos1988 committed:
updated remote card buttons layout

nikos1988 committed:
simplified remote card buttons layout

nikos1988 committed:
refactoring

zehnm commented:
@nikos1988 what's the state of this PR? It's still in draft and hasn't seen activity in a while.
Note: with the soon to be merged PR #522 there might be some changes required.

nikos1988 commented:
> 
> 
> @nikos1988 what's the state of this PR? It's still in draft and hasn't seen activity in a while.
> Note: with the soon to be merged PR #522 there might be some changes required.

Hi, i made some changes as Marton suggested. Actually i was waiting for a reply.

zehnm commented:
> Hi, i made some changes as Marton suggested. Actually i was waiting for a reply.

Ok, I'm changing it to "ready for review" so we can continue reviewing it.

zehnm committed:
Merge branch 'master' into feature/custom_buttons_remote

zehnm commented on config-schema.json:
```diff
@@ -455,6 +455,21 @@
+                  "examples": [
+                    "Button1, Button2, Button3"
+                  ],
+                  "pattern": ""
```
Either define a pattern or remove it. An empty pattern adds no value.
(In the next cleanup, I'll remove the other empty patterns in the schema).

nikos1988 replied:
I will remove the empty pattern

zehnm commented on config-schema.json:
```diff
@@ -455,6 +455,21 @@
+                "items": {
+                  "$id": "#/properties/entities/properties/remote/items/properties/custom_features/items",
+                  "type": "string",
+                  "title": "The Custom Features Schema",
```
Please add a descriptive title that matches the item.
This can improve validation error messages and allows us to create auto-generated edit forms in the future.

nikos1988 replied:
OK, i just imitated the other titles :)

zehnm commented on components/remote/ui/ir/IconButton.qml:
```diff
@@ -0,0 +1,63 @@
+/******************************************************************************
+ *
+ * Copyright (C) 2018-2019 Marton Borzak <hello@martonborzak.com>
```
I don't think Marton wrote that code

zehnm commented on components/remote/ui/ir/CardButtonsFunction.qml:
```diff
@@ -0,0 +1,112 @@
+/******************************************************************************
+ *
+ * Copyright (C) 2020 Marton Borzak <hello@martonborzak.com>
```
I don't think Marton wrote that code

zehnm commented on components/remote/ui/ir/CardButtonsCustom.qml:
```diff
@@ -0,0 +1,65 @@
+/******************************************************************************
+ *
+ * Copyright (C) 2020 Marton Borzak <hello@martonborzak.com>
```
I don't think Marton wrote that code

zehnm commented on components/remote/ui/ir/CardButtonsCustom.qml:
```diff
@@ -0,0 +1,65 @@
+            }
+        }
+
+        Component.onCompleted: {
+            //console.log("customButtonFlick.contentHeight:", customButtonFlick.contentHeight)
+            //console.log("customButtonRep.count:", customButtonRep.count)
+        }
```
Please no dead code

zehnm commented on components/remote/ui/ir/CardButtons.qml:
```diff
@@ -31,124 +32,141 @@ Item {
+
+            Button {
+                visible: obj.isSupported(Remote.F_DIGIT_ENTER)
+                title: "E"
```
Is this rendered on the button?
If yes, I think this should be translated. @martonborzak please verify

nikos1988 replied:
I was a bit confused about this feature. It was not used before so i made a placeholder.
The same applies for the DIGIT_SEPERATOR feature. There are no corresponding functions in the remote.cpp.

martonborzak replied:
This supposed to be the `enter` key on a keypad. I think these should not be translated as they are to mimic the "real" keys. So I would just call it "ENTER".

nikos1988 replied:
I don't think ENTER will fit the button since there are 3 buttons in a row

miloit replied:
Or "ENT" this is on my Samsung remote written

zehnm replied:
There are all kind of labels: "E", "ENT", "OK", graphical ones...  For space reasons I'd use "OK" or "E" and still put it in translations.
I guess @martonborzak needs to decide ;-)

martonborzak replied:
I vote for "ENT" as I think that is the closest to "ENTER".

zehnm commented on components/ButtonBase.qml:
```diff
@@ -49,7 +49,7 @@ Rectangle {
                 buttonContainer.opacity = 0.3
                 buttonContainer.enabled = false
-                buttonContainer.close();
+                //buttonContainer.close();
```
Why is close() no longer required?

nikos1988 replied:
Hi, the problem is that it is really annoying when you start a movie with the remote and want to pause it after 5 minutes.

When the remote wakes up from standby the entity is closed because it is not connected anymore and you have to reopen it again, go to the transport view and then press pause.
It is really inconvenient for a remote.

martonborzak replied:
That makes sense.

nikos1988 replied:
I will remove it completely then instead of commenting it

zehnm reviewed (changes requested):
- Please update the dependencies.cfg and set `integrations.library: v0.6.0` to make it compile.
- There were merge conflicts due to changes in the intgrations.library. Please verify that I've fixed them correctly, especially the cursor commands.
- @martonborzak what about the icons vs fonts? Personally I can't contribute there.
   Are svg a big overhead? We could check with the recently added fps counter...

nikos1988 committed:
cosmetic changes and copyright updates

zehnm committed:
Merge branch 'master' into feature/custom_buttons_remote

nikos1988 commented:
Okay, I just pushed an update

zehnm commented:
> Okay, I just pushed an update

Please update the dependencies.cfg and set `integrations.library: v0.6.0` to fix the compile error

zehnm commented on components/remote/ui/ir/CardButtonsCustom.qml:
```diff
@@ -0,0 +1,59 @@
+                Button {
+                    visible: true
+                    width: parent.width
+                    title: qsTr(modelData.replace(new RegExp("_", "g"), ' ')) + translateHandler.emptyString
```
@martonborzak could you check the qsTr? Afaik this only works for literal strings, not for dynamic content.

martonborzak replied:
Unfortunately,it's not gonna work like this.

nikos1988 replied:
So, what can we do about it?

martonborzak replied:
This is a custom button, so I assume the user will name it. Then simply just remove `qsTr` and `+ translateHandler.emptyString`

nikos1988 commented:
> 
> 
> > Okay, I just pushed an update
> 
> Please update the dependencies.cfg and set `integrations.library: v0.6.0` to fix the compile error

It is already on v0.6.0, isn't it?

zehnm commented:
> > > Okay, I just pushed an update
> > 
> > 
> > Please update the dependencies.cfg and set `integrations.library: v0.6.0` to fix the compile error
> 
> It is already on v0.6.0, isn't it?

Nope, the build is failing because of missing definitions in integrations.library. The `dependencies.cfg` is also missing in this PR. Update your local file and include it in the PR. 
The PR must successfully compile for further review.

nikos1988 committed:
removed translation for custom buttons

nikos1988 committed:
updated dependencies.cfg

nikos1988 commented:
> 
> 
> > > > Okay, I just pushed an update
> > > 
> > > 
> > > Please update the dependencies.cfg and set `integrations.library: v0.6.0` to fix the compile error
> > 
> > 
> > It is already on v0.6.0, isn't it?
> 
> Nope, the build is failing because of missing definitions in integrations.library. The `dependencies.cfg` is also missing in this PR. Update your local file and include it in the PR.
> The PR must successfully compile for further review.

v0.6.0 does not include the changes needed for the build to finish as far as i see it

zehnm commented:
It doesn't include the changes from your intg.lib PR yet, but other required changes.
I'll create a new intg.lib release.

zehnm commented on dependencies.cfg:
```diff
@@ -1,6 +1 @@
-#   name: dependency name
-#   separator: colon + optional white space
-#   version: git branch or tag
 integrations.library: v0.6.0
```
I've created a new release v0.7.0 which includes are changes in integrations.library.
Please update once more and don't delete the comments.

nikos1988 committed:
Revert "updated dependencies.cfg"

This reverts commit f68ef77acbc92c3964dcaad91e5492ed79b13508.

nikos1988 committed:
updated integrations.library to v0.7.0

nikos1988 commented:
OK, it builds fine now :)

zehnm reviewed (approved):
Looks good to me. Since it's mostly qml I'm handing over to @martonborzak

martonborzak commented on components/remote/ui/ir/Card.qml:
```diff
@@ -177,25 +290,72 @@ Rectangle {
-            asynchronous: true
-            sourceComponent: buttonView
+        Repeater {
+            model: pages()
```
This solution is really resource heavy. Let's leave it for now, but at some point it should be done by using a c++ `QAbstractListModel` to increase performance.

martonborzak commented on components/remote/ui/ir/CardButtonsFunction.qml:
```diff
@@ -0,0 +1,112 @@
+                visible: obj.isSupported(Remote.F_FUNCTION_RED)
+                width: height
+                border { color: Style.color.dark; width: 2 }
+                color: "#FF0000"
```
There's `Style.color.red`, please use that so we don't hardcode colors.

martonborzak commented on components/remote/ui/ir/CardButtonsFunction.qml:
```diff
@@ -0,0 +1,112 @@
+                color: "#FF0000"
+                mouseArea.onClicked: { obj.red(); }
+                mouseArea.onPressed: {}
+                mouseArea.onReleased: { redButton.color = "#FF0000" }
+                mouseArea.onCanceled: { redButton.color = "#FF0000" }
```
Same here.

martonborzak commented on components/remote/ui/ir/CardButtonsFunction.qml:
```diff
@@ -0,0 +1,112 @@
+                visible: obj.isSupported(Remote.F_FUNCTION_GREEN)
+                width: height
+                border { color: Style.color.dark; width: 2 }
+                color: "#008000"
```
There's `Style.color.green`, please use that so we don't hardcode colors.

martonborzak commented on components/remote/ui/ir/CardButtonsFunction.qml:
```diff
@@ -0,0 +1,112 @@
+                color: "#008000"
+                mouseArea.onClicked: { obj.green(); }
+                mouseArea.onPressed: {}
+                mouseArea.onReleased: { greenButton.color = "#008000" }
+                mouseArea.onCanceled: { greenButton.color = "#008000" }
```
Same here.

martonborzak commented on components/remote/ui/ir/CardButtonsFunction.qml:
```diff
@@ -0,0 +1,112 @@
+                visible: obj.isSupported(Remote.F_FUNCTION_YELLOW)
+                width: height
+                border { color: Style.color.dark; width: 2 }
+                color: "#FFFF00"
```
Please add yellow color to `Style.qml` and reference it here.

martonborzak commented on components/remote/ui/ir/CardButtonsFunction.qml:
```diff
@@ -0,0 +1,112 @@
+                mouseArea.onClicked: { obj.yellow(); }
+                mouseArea.onPressed: {}
+                mouseArea.onReleased: { yellowButton.color = "#FFFF00" }
+                mouseArea.onCanceled: { yellowButton.color = "#FFFF00" }
```
Same here.

martonborzak commented on components/remote/ui/ir/CardButtonsFunction.qml:
```diff
@@ -0,0 +1,112 @@
+                visible: obj.isSupported(Remote.F_FUNCTION_BLUE)
+                width: height
+                border { color: Style.color.dark; width: 2 }
+                color: "#0000FF"
```
There's `Style.color.blue`, please use that so we don't hardcode colors.

martonborzak commented on components/remote/ui/ir/CardButtonsFunction.qml:
```diff
@@ -0,0 +1,112 @@
+                color: "#0000FF"
+                mouseArea.onClicked: { obj.blue(); }
+                mouseArea.onPressed: {}
+                mouseArea.onReleased: { blueButton.color = "#0000FF" }
+                mouseArea.onCanceled: { blueButton.color = "#0000FF" }
```
Same here.

martonborzak commented on components/remote/ui/ir/CardButtonsFunction.qml:
```diff
@@ -0,0 +1,112 @@
+                visible: obj.isSupported(Remote.F_FUNCTION_ORANGE)
+                width: height
+                border { color: Style.color.dark; width: 2 }
+                color: "#FFA500"
```
There's `Style.color.orange`, please use that so we don't hardcode colors.

martonborzak commented on components/remote/ui/ir/CardButtonsFunction.qml:
```diff
@@ -0,0 +1,112 @@
+                color: "#FFA500"
+                mouseArea.onClicked: { obj.orange(); }
+                mouseArea.onPressed: {}
+                mouseArea.onReleased: { orangeButton.color = "#FFA500" }
+                mouseArea.onCanceled: { orangeButton.color = "#FFA500" }
```
Same here.

martonborzak commented on components/remote/ui/ir/CardButtonsTransport.qml:
```diff
@@ -29,91 +31,118 @@ Item {
+            spacing: 32
+            IconButton {
+                visible: obj.isSupported(Remote.F_RECORD)
+                image.source: "qrc:/images/remote/fiber_manual_record-white-48dp.svg"
```
For icons, best would be to extend the `icons.tft` font file with new icons and reference them from `Style.qml`.

martonborzak commented on components/remote/ui/ir/CardButtonsTransport.qml:
```diff
@@ -29,91 +31,118 @@ Item {
+
+            IconButton {
+                visible: obj.isSupported(Remote.F_PLAY)
+                image.source: "qrc:/images/remote/play_arrow-white-48dp.svg"
```
There's an icon defined in `Style.qml` `Style.icon.play`.

martonborzak commented on components/remote/ui/ir/CardButtonsTransport.qml:
```diff
@@ -29,91 +31,118 @@ Item {
+            IconButton {
+                visible: obj.isSupported(Remote.F_STOP)
+                image.visible: true
+                image.source: "qrc:/images/remote/stop-white-48dp.svg"
```
There's an icon defined in `Style.qml` `Style.icon.stop`.

martonborzak commented on components/remote/ui/ir/CardButtonsTransport.qml:
```diff
@@ -29,91 +31,118 @@ Item {
+            spacing: 32
+            IconButton {
+                visible: obj.isSupported(Remote.F_BACKWARD)
+                image.source: "qrc:/images/remote/fast_rewind-white-48dp.svg"
```
There's an icon defined in `Style.qml` `Style.icon.prev`.

martonborzak commented on components/remote/ui/ir/CardButtonsTransport.qml:
```diff
@@ -29,91 +31,118 @@ Item {
+            IconButton {
+                visible: obj.isSupported(Remote.F_PAUSE)
+                image.visible: true
+                image.source: "qrc:/images/remote/pause-white-48dp.svg"
```
There's an icon defined in `Style.qml` `Style.icon.pause`.

martonborzak commented on components/remote/ui/ir/CardButtonsTransport.qml:
```diff
@@ -29,91 +31,118 @@ Item {
+
+            IconButton {
+                visible: obj.isSupported(Remote.F_FORWARD)
+                image.source: "qrc:/images/remote/fast_forward-white-48dp.svg"
```
There's an icon defined in `Style.qml` `Style.icon.next`.

martonborzak commented on components/remote/ui/ir/CardButtonsTransport.qml:
```diff
@@ -29,91 +31,118 @@ Item {
+            spacing: 32
+            IconButton {
+                visible: obj.isSupported(Remote.F_PREVIOUS)
+                image.source: "qrc:/images/remote/skip_previous-white-48dp.svg"
```
There's an icon defined in `Style.qml` `Style.icon.prev`.

martonborzak commented on components/remote/ui/ir/CardButtonsTransport.qml:
```diff
@@ -29,91 +31,118 @@ Item {
+
+            IconButton {
+                visible: obj.isSupported(Remote.F_INFO)
+                image.source: "qrc:/images/remote/info-white-48dp.svg"
```
For icons, best would be to extend the icons.tft font file with new icons and reference them from Style.qml.

martonborzak commented on components/remote/ui/ir/CardButtonsTransport.qml:
```diff
@@ -29,91 +31,118 @@ Item {
+
+            IconButton {
+                visible: obj.isSupported(Remote.F_NEXT)
+                image.source: "qrc:/images/remote/skip_next-white-48dp.svg"
```
There's an icon defined in `Style.qml` `Style.icon.next`.

martonborzak commented on components/remote/ui/ir/CardButtonsTransport.qml:
```diff
@@ -29,91 +31,118 @@ Item {
+
+            IconButton {
+                visible: obj.isSupported(Remote.F_HOME)
+                image.source: "qrc:/images/remote/home-white-48dp.svg"
```
There's an icon defined in `Style.qml` `Style.icon.home`.

martonborzak commented on components/remote/ui/ir/CardButtonsTransport.qml:
```diff
@@ -29,91 +31,118 @@ Item {
+
+            IconButton {
+                visible: obj.isSupported(Remote.F_MENU)
+                image.source: "qrc:/images/remote/menu-white-48dp.svg"
```
For icons, best would be to extend the icons.tft font file with new icons and reference them from Style.qml.

nikos1988 committed:
use Style colors for buttons

nikos1988 committed:
use icon font for buttons with icons

martonborzak committed:
Add missing icons

nikos1988 committed:
use newly added icons

nikos1988 commented:
OK, icons should be all okay now

martonborzak reviewed (approved):
Thanks, looks good!

zehnm committed:
Merge branch 'master' into feature/custom_buttons_remote
